velox/core/QueryCtx.h

```diff
@@ -28,25 +28,13 @@ namespace facebook::velox::core {
 
 class QueryCtx : public Context {
  public:
-  // Returns QueryCtx with new executor created if not supplied. For testing
-  // purpose.
-  static std::shared_ptr<QueryCtx> createForTest(
-      std::shared_ptr<Config> config = std::make_shared<MemConfig>(),
-      std::shared_ptr<folly::Executor> executor =
-          std::make_shared<folly::CPUThreadPoolExecutor>(
-              std::thread::hardware_concurrency())) {
-    return std::make_shared<QueryCtx>(std::move(executor), std::move(config));
-  }
-
   // QueryCtx is used in different places. When used with `Task`, it's required
   // that callers supply executors. In contrast, when used in expression
   // evaluation through `ExecCtx`, executor is not needed. Hence, we don't
   // require executor to always be passed in here, but instead, ensure that
   // executor exists when actually being used.
-  //
-  // This constructor keeps the ownership of `executor`.
   QueryCtx(
-      std::shared_ptr<folly::Executor> executor = nullptr,
+      folly::Executor* FOLLY_NULLABLE executor = nullptr,
```

Comment:
Let's update the comments to explain that the caller is responsible for ensuring the lifetime of the Executor to exceed the lifetime of all the tasks that use it.

> When used with `Task`, it's required that callers supply executors.

Perhaps, clarify that this is the case only if the caller is using Task::start(). The executor is not required when using Task::next().

Reply:
Thanks for pointing out!